src/bevent.c

```diff
@@ -410,6 +752,168 @@ void module_event(const char *module, const char *event, struct ua *ua,
 }
 
 
+static void bevent_emit_base(struct bevent *event)
+{
+	struct le *le;
+	le = ehel.head;
+	while (le) {
+		struct ehe *ehe = le->data;
+		le = le->next;
+
+		ehe->h(event->ev, event, ehe->arg);
+
+		if (event->stop)
+			return;
+	}
+}
+
+
+static int bevent_emit(struct bevent *event, const char *fmt, va_list ap)
+{
+	char *buf;
+	int err;
+
+	if (!fmt)
+		err = re_sdprintf(&buf, "");
+	else
+		err = re_vsdprintf(&buf, fmt, ap);
+	if (err)
+		return err;
+
+	event->txt = buf;
+	event->err = 0;
+	bevent_emit_base(event);
+	if (event->err) {
+		err = event->err;
+		goto out;
+	}
+
+	if (event->stop)
+		goto out;
+
+out:
+	mem_deref(buf);
+	return err;
+}
+
+
+/**
+ * Emit an application event
+ *
+ * @param ev   User-agent event
+ * @param arg  Application specific argument (optional)
+ * @param fmt  Formatted arguments
+ * @param ...  Variable arguments
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int bevent_app_emit(enum ua_event ev, void *arg, const char *fmt, ...)
```

Comment:
What is the advantage to having different API functions for the different event classes? They all pass pointers as arguments anyways, wouldn't a single API function `bevent_emit` be enough?

Reply:
Different event classes use different union fields. 
And the different emit functions have different parameters. This is the major improvement.

E.g. If the user wants to emit an application event, he doesn't want to take care about a call object.

Reply:
Also it is very easy to add new event classes, without breaking the backwards compatibility.

Reply:
Okay, so the advantage I see is that if the API is used correctly, type checking is enforced and it is more difficult to incorrectly cast some void pointers to a wrong type. Sounds good to me.